### PR TITLE
ImplTreeSet

### PR DESCRIPTION
**Pull request by xRuBYN**

### CODE REVIEW

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 1]
```diff
@@ -1,84 +1,164 @@
...
+        return tree.getSize();
     }
 
     @Override
     public boolean isEmpty() {
-        //TODO
+        if (tree.head == null) {
+            return true;
+        }
         return false;
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/collections/StudentSet.java]:
```suggestion
        return tree.head == null;
```

File: src/main/java/com/endava/internship/collections/Node.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.endava.internship.collections;
+
+public class Node {
+    //Node head;
+    Node left;
+    Node right;
+    Student value;
+
+    public Node() {
+    }
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/collections/Node.java]: can remove unused constructor

File: src/main/java/com/endava/internship/collections/Node.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package com.endava.internship.collections;
+
+public class Node {
+    //Node head;
+    Node left;
+    Node right;
+    Student value;
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/collections/Node.java]: instance variables must be private

File: src/main/java/com/endava/internship/collections/BTree.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    public boolean searchElement(Node current, Student value) {
+        if (current == null) {
+            return false;
+        }
+        if (current.value.compareTo(value) == 0) {
+            return true;
+        }
+        return (current.value.compareTo(value) > 0 ? true : false)
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/collections/BTree.java]:
```suggestion
        return current.value.compareTo(value) > 0
```

File: src/main/java/com/endava/internship/collections/BTree.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        } else if (root.value.compareTo(value) < 0) {
+            root.right = insert(root.right, value);
+
+        } else if (root.value.compareTo(value) > 0) {
+
+            root.left = insert(root.left, value);
+        }
+        return root;
+
+    }
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/collections/BTree.java]: please reorder methods
https://www.oracle.com/java/technologies/javase/codeconventions-fileorganization.html
image

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 6]
```diff
@@ -1,84 +1,164 @@
...
+                check = true;
+            }
+        }
+        return check;
+    }
+
+    @Override
+    public String toString() {
+        tree.traverseInOrder(tree.head);
+        return " ";
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/collections/StudentSet.java]: we should not return space))

File: src/main/java/com/endava/internship/collections/Student.java [hunk 7]
```diff
@@ -1,11 +1,13 @@
 package com.endava.internship.collections;
 
 import java.time.LocalDate;
+import java.util.Comparator;
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/collections/Student.java]: remove unused import please

File: src/main/java/com/endava/internship/collections/Student.java [hunk 8]
```diff
@@ -17,16 +19,63 @@ public Student(String name, LocalDate dateOfBirth, String details) {
...
+    //if this element more than compare element return positive value
+    @Override
+    public int compareTo(Student o) {
+        if (0 == this.name.compareTo(o.name)) {
+            if (this.dateOfBirth.compareTo(o.dateOfBirth) < 0) {
+                return -1;
+            } else if (this.dateOfBirth.compareTo(o.dateOfBirth) > 0) {
+                return 1;
+            } else
+                return 0;
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/collections/Student.java]:
```suggestion
            return Integer.compare(this.dateOfBirth.compareTo(o.dateOfBirth), 0);
```

File: src/main/java/com/endava/internship/collections/StudentSet.java (hunk 1 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/collections/StudentSet.java]: fixed

File: src/main/java/com/endava/internship/collections/Node.java (hunk 2 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/collections/Node.java]: fixed

File: src/main/java/com/endava/internship/collections/Node.java (hunk 3 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/collections/Node.java]: fixed

File: src/main/java/com/endava/internship/collections/BTree.java (hunk 4 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/collections/BTree.java]: fixed

File: src/main/java/com/endava/internship/collections/BTree.java (hunk 5 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/collections/BTree.java]: fixed

File: src/main/java/com/endava/internship/collections/Student.java (hunk 7 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/collections/Student.java]: fixed

File: src/main/java/com/endava/internship/collections/Student.java (hunk 8 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/collections/Student.java]: fixed

File: src/main/java/com/endava/internship/collections/StudentSet.java (hunk 6 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/collections/StudentSet.java]: fixed

File: src/main/java/com/endava/internship/collections/Student.java [hunk 9]
```diff
@@ -17,16 +19,50 @@ public Student(String name, LocalDate dateOfBirth, String details) {
...
 
-    /*
-    TODO consider overriding any methods for this object to function properly within a collection:
-        1. A student is considered unique by a combination of their name and dateOfBirth
-        2. Student names are sorted alphabetically, if two students have the same name, then the older one is
-        placed before the younger student in an ordered student list.
-    */
+    //if this element equal with compare element return 0
+    //if this element less than compare element return negative value
+    //if this element more than compare element return positive value
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/collections/Student.java]: You can use javadoc comment here

File: src/main/java/com/endava/internship/collections/BTree.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        return root;
+    }
+
+    public StringBuilder traverseInOrder(Node tree) {
+        if (tree != null) {
+            traverseInOrder(tree.getLeft());
+            str.append(tree.getValue()).append("\n");
+            traverseInOrder(tree.getRight());
+        }
+        return str;
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/collections/BTree.java]: after execution of this method, we don't need to do other manipulation on StringBuilder, so we can return a String,

**Review by meedvv379 — APPROVED** [review]

File: src/main/java/com/endava/internship/collections/BTree.java (hunk 10 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/collections/BTree.java]: fixed

File: src/main/java/com/endava/internship/collections/Student.java (hunk 9 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/collections/Student.java]: fixed

**Pull request merged**